### PR TITLE
`Discussion`: Improve usability in discussion overview

### PR DESCRIPTION
**Pull request by lschlesinger**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the test guidelines.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Some things were on my To-Do-List:
- Remove `votes` from the `Post` class, as the data model changed and voting is reflected by `reactions` (the data column will be dropped in a follow-up to ensure compatibility of DB's on test servers)

- Renaming: `PostContentCompareStrategy`-> `PostSimilarityComparisonStrategy`, as the check does not necessarily depends on the content but also take other `Post` Metadata in account when checking for similarity
- Renaming: `PageDiscussionSectionComponent` -> `DiscussionSectionComponent`, no need to have `Page` in the naming
- Renaming: all components that include both types, `Post` and `AnswerPost` had `Posting` (superclass) in the name, but with inconsistent convention `postings-xy` vs. `posting-xy`, now changed to `posting-xy`

- Discussion Overview: options that can be selected as context, i.e., lectures and exercises are sorted alphabetically to enable users to quickly find the context they want to select
- Discussion Overview: when a user started to create a post, cancelled it (clicking x) and then opened the modal again, the context selection was not properly reset (first input field of the modal), it should be `Tech Support` on default
- Discussion Overview: Announcement (pinned and highlighted posts) were not listed on top of the list if there are other posts that were created afterwards and also pinned, because the creation date was the second criterion to order posts after the display priority criterion (pinned, archived). I introduced the announcement type criterion as first order criterion now. 

### Description

All those changes are not complex to understand, most important is, that it does not breaking a working UI.
The tests I had to touch were also updated in accordance with our newest client test guidelines.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Course with posts enabled

1. Log in to Artemis
2. Navigate to discussion overview
3. As instructor start to create an announcement (select Announcement as context), and cancel before saving, i.e. click `x` at the top right of the modal
4. Start to create a new post again by clicking the button `New Post` -> check that the context is set to Tech Support on start-up
5. Select announcement as context -> check if the blue info occurs right below the context
6. Select any other context and provide a title -> check that similar posts are shown below in a list
7. Save the post
8. Interact with existing posts -> editing, deleting, react, answering -> check that no unexpected behavior occurs
9. Create an announcement and another post afterwards. Pin the other post -> the announcement should be listed at top (even though the second post was created more recently and is also pinned)
10. Remove the pin of the announcement -> it should be listed after it's creation date
11. Archive the announcement -> it should be moved to the bottom of the list
12. Log in as student and interact with a posting -> create, edit own, delete own, answer -> check that no unexpected behavior occurs

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

Instructor who is checking the contexts which are sorted alphabetically and cancels + creates a post as described in the testing steps.

https://user-images.githubusercontent.com/41366522/142689015-85af9d79-b452-4cfd-9704-31fc82ea1b63.mov

### CODE REVIEW

**Review by VinnavanSembian — APPROVED** [review]
Tested in testing session. Everything works as it should.

**Review by mallgaier — APPROVED** [review]
Pull Request tested according to the test steps. Behavior was as expected; without unexpected error messages.

**Review by lBruchmann — APPROVED** [review]
Tested during testing session. Tested on ts3. I followed the testing steps. Everything works as expected. Creating announcements, editing and deleting posts works without error messages.

**Review by cremertim — DISMISSED** [review]
i tested all functionallities described in the testing steps by both sides, the students and the instructor during the testing session. I did not encounter any UI breaking Issues and everything worked as before. 

Nice refactoring 👍

File: src/main/webapp/app/shared/metis/post/post.component.html [hunk 1]
```diff
@@ -25,9 +25,8 @@
                     [routerLink]="contextInformation.routerLinkComponents"
                     (click)="onNavigateToContext($event)"
                     routerLinkActive="active"
+                    >{{ contextInformation.displayName + ':' }}</a
                 >
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/post/post.component.html:29]:
```suggestion
                    >
                        {{ contextInformation.displayName + ':' }}
                    </a>
```
This is the code style that was also used below in this file (lines 40 -42). We should be consistent here, the suggested one looks better imo.

Edit: Just noticed that this is exectly how the file looked before. What is the reason behind this change?

File: src/main/webapp/app/shared/metis/posting-create-edit-modal/post-create-edit-modal/post-create-edit-modal.component.ts [hunk 2]
```diff
@@ -65,36 +62,44 @@ export class PostCreateEditModalComponent extends PostingsCreateEditModalDirecti
...
     ngOnChanges(): void {
-        this.resetCurrentContextSelectorOption();
         super.ngOnChanges();
     }
 
     /**
      * resets the pageType, initialContext, post tags, post title, and post content
      */
     resetFormGroup(): void {
+        console.log('reset formgroup');
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/metis/posting-create-edit-modal/post-create-edit-modal/post-create-edit-modal.component.ts:72]:
```suggestion
```
Did you add the log on purpose here?

**Review by Strohgelaender — DISMISSED** [review]
Tested in testing session, we could not find any issues. I only added two small questions about the code changes.

File: src/main/webapp/app/shared/metis/post/post.component.html (hunk 1 above)
Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/post/post.component.html:29]: The difference is if the whitespace between `:` and the title is included in the link.
Current solution:

Bildschirmfoto 2021-11-23 um 16 09 23

Suggested solution:
Bildschirmfoto 2021-11-23 um 16 09 59

I want the current version, but I will move the entire tag into line 28. It was not intended that `>` is in a new line.

File: src/main/webapp/app/shared/metis/posting-create-edit-modal/post-create-edit-modal/post-create-edit-modal.component.ts (hunk 2 above)
Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/posting-create-edit-modal/post-create-edit-modal/post-create-edit-modal.component.ts:72]: nope, I remove it. thanks!

File: src/main/webapp/app/shared/metis/post/post.component.html (hunk 1 above)
Comment (julian-christl) [reply src/main/webapp/app/shared/metis/post/post.component.html:29]: This is btw exactly the case where HTML renders one or multiple newlines as one space character.

You could move the colon out of the Angular brackets btw. Not sure if you agree with that style change, please see how it looks codewise.

Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/post/post.component.html]: @julian-christl thanks for the explanation, do you agree with the changes in d89cc69?

Comment (julian-christl) [reply src/main/webapp/app/shared/metis/post/post.component.html]: Yes, that's what I meant with the colon.

inlining the end tags is not possible with Angular, is it? (e.g. line 39 and 40)

Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/post/post.component.html]: nope, I tried it several times, the linter always changed it. I just recognized that, after I got very confused :D

Comment (julian-christl) [reply src/main/webapp/app/shared/metis/post/post.component.html]: Sounds good to me. I think the linter doesn't like it because the starting tag is not inline.
